matsim/src/main/java/org/matsim/core/population/io/ParallelPopulationReaderMatsimV6.java

```diff
@@ -0,0 +1,222 @@
+
+	static final Logger log = LogManager.getLogger(ParallelPopulationReaderMatsimV6.class);
+
+	private final boolean isPopulationStreaming;
+	private final int numThreads;
+	private final BlockingQueue<List<Tag>> queue;
+	private Thread[] threads;
+	private List<Tag> currentPersonXmlData;
+
+	private final String inputCRS;
+	private final String targetCRS;
+
+	private boolean reachedPersons = false;
+
+
+	public ParallelPopulationReaderMatsimV6(
+			final String inputCRS,
+			final String targetCRS,
+			final Scenario scenario) {
+		super(inputCRS, targetCRS, scenario);
+		this.inputCRS = inputCRS;
+		this.targetCRS = targetCRS;
+
+		/*
+		 * Check whether population streaming is activated
+		 */
+
+		if (scenario.getPopulation() instanceof StreamingPopulationReader.StreamingPopulation) {
+			log.warn("Population streaming is activated - cannot use " + ParallelPopulationReaderMatsimV6.class.getName() + "!");
+
+			this.isPopulationStreaming = true;
+			this.numThreads = 1;
+			this.queue = null;
+
+		} else {
+			isPopulationStreaming = false;
+
+			if (scenario.getConfig().global().getNumberOfThreads() > 0) {
+				this.numThreads = scenario.getConfig().global().getNumberOfThreads();
+			} else this.numThreads = 1;
+
+			this.queue = new LinkedBlockingQueue<>();
+		}
+	}
+
+	private void initThreads() {
+		threads = new Thread[numThreads];
+		for (int i = 0; i < numThreads; i++) {
+
+			ParallelPopulationReaderMatsimV6Runner runner =
+					new ParallelPopulationReaderMatsimV6Runner(
+							this.inputCRS,
+							this.targetCRS,
+							this.scenario,
+							this.queue);
+
+			Thread thread = new Thread(runner);
+			thread.setDaemon(true);
+			thread.setName(ParallelPopulationReaderMatsimV6Runner.class.toString() + i);
+			threads[i] = thread;
+			thread.start();
+		}
+	}
+
+	@Override
+	public void startTag(String name, Attributes atts, Stack<String> context) {
+		//Reached first time a person
+		if (PERSON.equals(name) && !this.reachedPersons) {
+			this.reachedPersons = true;
+
+			if (!isPopulationStreaming && this.threads == null) {
+				log.info("Start parallel population reading...");
+				initThreads();
+			}
+		}
+
+		// if population streaming is activated, use non-parallel reader
+		if (isPopulationStreaming || !this.reachedPersons) {
+			super.startTag(name, atts, context);
+			return;
+		}
+
+
+		// If it is a new person, create a new person and a list for its attributes.
+		if (PERSON.equals(name)) {
+			Person person = this.plans.getFactory().createPerson(Id.create(atts.getValue(ATTR_PERSON_ID), Person.class));
+			currentPersonXmlData = new ArrayList<>();
+			PersonTag personTag = new PersonTag();
+			personTag.person = person;
+			currentPersonXmlData.add(personTag);
+			this.plans.addPerson(person);
+		} else {
+
+			// Create a new start tag and add it to the person data.
+			Stack<String> contextCopy = new Stack<>();
+			contextCopy.addAll(context);
```

Comment:
This might be a costly operation, but there is no alternative except writing a custom Stack implementation and using it in the abstract MatsimXmlParser :-/

Reply:
Thats right, a copy is needed as this context is used by the parser, similar like the attributes.